transform_constantfold.ml

```diff
@@ -1,21 +1,89 @@
 open Instr
 
 (*
- * Constant propagation. Takes a program `prog` and returns an updated stream.
+ * Constant folding.
  *
- * Finds all constant declarations of the form:
- *     var x = l
- * where `l` is a literal.
+ * This optimization combines constant propagation and constant folding.
+ *
+ * A variable is constant if
+ *   (1) It is defined as `var x = l` where `l` is a literal
+ *   (2) It is defined as `var x = op(se)` where `op(se)` is a primop involving
+ *       literals or constant variables
  *
  * Then, whenever `x` is used in an expression (while x is still in scope), it
- * is replaced by the literal `l`. Afterwards, the variable `x` is no longer
+ * is replaced by its constant value. Afterwards, the variable `x` is no longer
  * used, and the declaration can be removed by running `minimize_lifetimes`.
  *)
-let const_prop ({formals; instrs} : analysis_input) : instructions option =
-  (* Replaces the variable `x` with literal `l` in instruction `instr`. *)
-  let replace x l instr =
-    let replace = Edit.replace_var_in_exp x l in
-    let replace_arg = Edit.replace_var_in_arg x l in
+let const_fold ({formals; instrs} : analysis_input) : instructions option =
+  let module VarMap = Map.Make(Variable) in
+  let module Approx = struct
+    type t = Unknown | Value of value
+    let equal a1 a2 = match a1, a2 with
+      | Unknown, Unknown -> true
+      | Unknown, Value _ | Value _, Unknown -> false
+      | Value v1, Value v2 -> Eval.value_eq v1 v2
+  end in
+
+  (* Constant fold an expression. *)
+  let fold exp =
```

Comment:
proposal: just call `Eval.eval` instead, after checking that `exp_vars exp` is empty.